### PR TITLE
Rename `coreos.firstboot` kernel cmdline arg to `ignition.firstboot`

### PR DESCRIPTION
**Pull request by LorbusChris**
to make options more distribution neutral.
Update dracut/README.txt to take into account recent changes.

Resolves:  #24 #29

### CODE REVIEW

File: grub/02_ignition_firstboot [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
+exec tail -n +3 $0
+set ignition_firstboot=""
+# Determine if this is a first boot.
+if [ -f "/ignition-firstboot" ]; then
+    set ignition_firstboot="ignition.firstboot=detected"
```
Comment (dustymabe) [inline grub/02_ignition_firstboot]: so I don't think we actually need the `=detected` part. can we try it just with `ignition.firstboot` by itself?

**Comment by dustymabe** [comment]
this doesn't resolve #29 I don't think so we should probably remove that from the commit message.

File: grub/02_ignition_firstboot (hunk 1 above)
Comment (cgwalters) [reply grub/02_ignition_firstboot]: The `=detected` was there for a reason...it relates to the diskless PXE case I think.  @bgilbert would know.

**Comment by cgwalters** [comment]
So umm...I know that we don't want to carry a legacy forever but there's so many git repositories and builds involved here.  And the failure case is going to be Ignition not running which is going to be painful.

How about we support both for say a month, and the follow up again and drop the old one?

**Comment by dustymabe** [comment]
> How about we support both for say a month, and the follow up again and drop the old one?

I originally started to agree with you here, but I think it might be more complicated than I originally thought. What does "support both" mean to you? I think we'd have to do more than just supporting the two kernel args.

File: grub/02_ignition_firstboot (hunk 1 above)
Comment (LorbusChris) [reply grub/02_ignition_firstboot]: I've removed it for now, let me know if we need it

**Comment by cgwalters** [comment]
> What does "support both" mean to you? I think we'd have to do more than just supporting the two kernel args.

Yeah, we'd need to assume firstboot if we see either one, and *also* delete both files on completion or something?

File: grub/02_ignition_firstboot (hunk 1 above)
Comment (bgilbert) [reply grub/02_ignition_firstboot]: In CL the difference is: PXE sets `coreos.first_boot=1`, which runs Ignition.  GRUB sets `coreos.first_boot=detected`, which runs Ignition and then mounts `/boot` to delete the flag file.  If we didn't distinguish between those, PXE would wait forever for the (nonexistent) `/boot` filesystem to show up.

Comment (dustymabe) [reply grub/02_ignition_firstboot]: in FCOS right now I think we are using `ignition-firstboot-complete.service` to remove the file so I think we don't have that same problem?

Comment (bgilbert) [reply grub/02_ignition_firstboot]: As of this PR, we still do:

```
ConditionKernelCommandLine=ignition.firstboot
RequiresMountsFor=/boot
```

Dropping the `RequiresMountsFor` would create a race in the other direction, if `/boot` is slow to mount.  It might be possible to condition on PXE rather than on a distinct firstboot cmdline value, if we have a good way to detect PXE.

Note that conditionalization can't happen in `ignition-firstboot-complete.service` because systemd would still unconditionally pull in the unit's dependencies, including `/boot`.  The service will need to be conditionally enabled by a generator instead.

**Comment by dustymabe** [comment]
> Yeah, we'd need to assume firstboot if we see either one, and _also_ delete both files on completion or something?

talked with colin.. we're going to coordinate the change (i'll plan to do this on Monday) so that we don't have to support multiple options in multiple places. Thanks colin for discussing that with me and developing a plan together.

File: dracut/30ignition/ignition-generator [hunk 2]
```diff
@@ -35,7 +35,7 @@ add_requires() {
 # This can't be done with ConditionKernelCommandLine because that always
 # starts the unit's dependencies. We want to start networkd only on first
 # boot.
-if $(cmdline_bool 'coreos.firstboot' 0); then
+if $(cmdline_bool 'ignition.firstboot' 0) && $(cmdline_bool 'coreos.firstboot' 0); then
```
Comment (bgilbert) [inline dracut/30ignition/ignition-generator]: It looks like this is requiring both arguments to be specified?

Comment (dustymabe) [reply dracut/30ignition/ignition-generator]: indeed.. so we just need to drop to the one `ignition.firstboot` check and delete the other one.

File: grub/02_ignition_firstboot (hunk 1 above)
Comment (LorbusChris) [reply grub/02_ignition_firstboot]: not too pretty, but maybe this would work:
`ignition-firstboot-non-pxe.service`:
```
[Unit]
ConditionKernelCommandLine=ignition.firstboot
ConditionPathIsMountPoint=/boot

[Service]
Type=oneshot
ExecStart=/bin/systemctl start ignition-firstboot-non-pxe-cleanup.service
```
`ignition-firstboot-non-pxe-cleanup.service`:
```
[Unit]
RequiresMountsFor=/boot

[Service]
Type=oneshot
RemainAfterExit=yes
ExecStart=/bin/rm /boot/ignition.firstboot
```

Comment (dustymabe) [reply grub/02_ignition_firstboot]: I've been thinking about this a bit more. I've never run a diskless PXE case (I've run PXE for installs, but not for diskless). I realize in the PXE case why we don't **need** a /boot/ partition (i.e. kernel and initrd are loaded via PXE over network), but why not have the disk image that gets mounted just be the same disk image that we serve everywhere else? i.e. so /boot/ does exist and the file can get deleted without issue (even though the delete would be pointless).

Comment (dustymabe) [reply grub/02_ignition_firstboot]: discussed a little bit in an open discussion today. i think for now we agreed to leave `ignition.firstboot=detected` out and add it back once we add in PXE support if we still want to implement it the same way it was done in the past for CL

Comment (bgilbert) [reply grub/02_ignition_firstboot]: @LorbusChris `ExecStart=/bin/systemctl start` is never a good idea, since it hides dependencies from systemd's dependency system.  Also, I believe `ConditionPathIsMountPoint` detects whether the specified directory is a mountpoint at this exact moment, so we'd still have the race.  A systemd generator is the right way to handle conditional unit enablement.

@dustymabe The PXE image would have the same files, but not necessarily the same partition layout.  systemd would need to know not to try to mount a dedicated partition on `/boot`.

After more thought, I'm not sure `RequiresMountsFor` is actually the problem.  It seems as though systemd in the real root should know whether the ESP should be mounted on `/boot`; however, we'll still need a generator to conditionalize that.

File: dracut/30ignition/ignition-generator (hunk 2 above)
Comment (LorbusChris) [reply dracut/30ignition/ignition-generator]: oh yea, this was a quick shot.

File: systemd/ignition-firstboot-complete.service [hunk 3]
```diff
@@ -2,18 +2,18 @@
 Description=Mark boot complete
 Documentation=https://github.com/coreos/ignition
 # Note that deriving from Container Linux, there is
-# a tradition that coreos.first_boot=1 is used for PXE/diskless
-# scenarios where it's *always* first boot, so we don't want
-# to try to change `/boot` there.
-ConditionKernelCommandLine=coreos.firstboot=detected
+# a tradition that coreos.first_boot=1 (now ignition.firstboot)
```
Comment (LorbusChris) [inline systemd/ignition-firstboot-complete.service]: Maybe rephrase this to clearify?

**Comment by LorbusChris** [comment]
Hold this until v3 has been branched?

File: systemd/ignition-firstboot-complete.service (hunk 3 above)
Comment (dustymabe) [reply systemd/ignition-firstboot-complete.service]: I think we can possibly just remove this entire comment for now until we "add back" diskless PXE support

**Comment by dustymabe** [comment]
> Hold this until v3 has been branched?

I don't think we need to do that. The only part that needs to wait for v3 is [the ignition PR](https://github.com/coreos/ignition/pull/663)

**Comment by dustymabe** [comment]
LGTM. @bgilbert can I get a final review from you?
